contributors/design-proposals/defer-containers.md

```diff
@@ -0,0 +1,248 @@
+Kubernetes.  Such stateful pods benefit from constructor and destructor semantics. InitContainers provide constructor
+semantics. The proposed DeferContainers will provide destructor semantics for such stateful PODs
+
+## Abstract
+This introduces the concept of ‘deferContainers’, inspired by golang’s keyword ‘defer’. This will allow one to define a
+set of Containers to be executed at the time of POD termination. The defined containers will be executed sequentially and
+in the specified order. It will bring destructor() capability to Pods.
+
+## Use Cases
+
+### Cleanup:
+Most stateful workloads require certain Cleanup activity before or after the appContainers are terminated such as
+* Sync/Flush the Disk or logs before the POD is evicted off a node.
+* Delete or Update Global Configuration / Release Application level Lock.
+* Some Legacy applications may not respond to SIGTERM signals sent by docker daemon, they might require special command/procedure
+  to initiate graceful shutdown.
+
+### Shard rebalancing.
+In a sharded workload configured as a statefulset, if the application is scaled down, then the exiting Pod should trigger
+a re-balance of keys from the current shard to the rest of them. This needs to be done befoe the shard goes down.  
+‘deferContainers’ could handle such scenarios with ease.
+
+### Statefulset Upgrading
+If in future, statefulsets support rolling updates, an update request might attempt to replace Pods one after the other. 
+With deferContainers each Pod could be removed gracefully without having to implement complicated sidecar logic to prevent data loss.
+
+### Master-Slave or Leader-Follower Statefulset down-size
+Consider a Stateful set consisting a Master-Slave or Leader-follower type application running N replicas. Ordinal index 
+ranging from 0 to N-1.  If you scale it down (reduce the number of replicas), the statefulset controller would attempt 
+to bring down the last replica (N-1th pod in this case).  If that pod is the elected Leader or Master, then this would 
+disrupt the service intermittently. With deferContainers, this handover or re-election could be programmed gracefully/elegantly.
+
+### SHUTDOWN Sequence
+* Traditional relational  databases typically support a reasonable shutdown sequence, for instance, Oracle has 4 types of 
+shutdown such as NORMAL, IMMEDIATE, TRANSACTIONAL and ABORT. ‘deferContainers’ will allow us to program and wait for such
+complex shutdown scenarios. 
+* In future when kuberenetes supports Virtual Machine runtime (eg: hyperContainer) for better isolation, we should shutdown
+  the VMs instead of killing them abruptly.  ‘deferContainers’ could help us run such shutdwon commands.
+
+## Limitation with the current system
+Container pre-stop hooks are not sufficient for all termination cases:
+* They cannot easily coordinate complex conditions across containers in multi-container pods
```

Comment:
The container hooks are not generic to POD itself they are specific to containers, imagine a multi-container pod with 1 App Container and 1 Side Car 
SideCar : Membership/Gossip Container
AppContainer: A Stateful Sharded workload
Termination Sequence: First one needs to re-shard/rebalance and then come out of membership
Below code suggests that both the preStop hooks (if configured for both the containers) will be started in parllel
```go
for _, container := range runningPod.Containers {
		go func(container *kubecontainer.Container) {
			defer utilruntime.HandleCrash()
			defer wg.Done()

			killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, container.Name)
			if err := m.killContainer(pod, container.ID, container.Name, "Need to kill Pod", gracePeriodOverride); err != nil {
				killContainerResult.Fail(kubecontainer.ErrKillContainer, err.Error())
			}
			containerResults <- killContainerResult
		}(container)
	}
```
1) hard to co-ordinate if both have different preStop hooks, 
2) hard consolidate this logic to one preStop script if both are from different images. Individually upgrading the images becomes a problem
3) If only one preStop command failed only that should be retried and not the whole script(currently preStop hooks restarts are not available user shoudl include them in the script)